documentation/ArrayExprLoweringToFIR.md

```diff
@@ -0,0 +1,303 @@
+### Alternative – Data Approach 
+The data-oriented approach materializes every array subexpression (but subsections and transformational that use `fir.box<>`). The FIR generated for the example 
+```
+// Allocate temporary storages on the stack
+%tmp1 = fir.alloca fir.array<n:m:fir.real<4>> 
+%tmp2 = fir.alloca fir.array<n:m:fir.real<4 >>
+
+// Create a box over A to implement A(2*m+1:4*m:2,:)
+%0 = fir.gendims 2*m+1, 4*m, 2
+%1 = fir.gendims 1, n, 1
+%asection = fir.embox %a, %0, %1
+
+// Compute A(2*m+1:4*m:2,:) * B into %tmp1
+fir.loop %i = 1 to m {
+  fir.loop %j = 1 to n {
+    %3 = fir.extract_value %b %i, %j
+    %4= fir.fmul %2 %3
+    %5= fir.coordinate_of %tmp1 %i %j
+    fir.store %4 to %5
+  }
+}
+
+// Compute foo(x)
+%6 = fir.call %x
+
+// Represent TRANSPOSE (C) mapping over C
+#transpose = (i, j)[] -> (j, i)
+%ct= fir.embox %c [#transpose]
+
+// Compute foo(x) * TRANSPOSE(C) into %tmp2
+fir.loop %i = 1 to m {
+  fir.loop %j = 1 to n {
+    %7 = fir.extract_value %ct %i, %j
+    %8= fir.fmul %6 %7
+    %9= fir.coordinate_of %tmp2 %i %j
+    fir.store %8 to %9
+  }
+}
+
+fir.loop %i = 1 to m {
+  fir.loop %j = 1 to n {
+    %10 = fir.extract_value %tmp1 %i, %j
+    %11 = fir.extract_value %tmp2 %i, %j
+    %12= fir.fadd %10 %11
+    %13= fir.coordinate_of %res %i %j
+    fir.store %12 to %13
+  }
+}
+```
+Its main advantage is that it is easy to generate in single bottom-up expression tree visit (generating FIR for an expression node at the post visit). No particular states need to be maintained regarding where the FIR code should be emitted (the `mlir::OpBuilder` is sufficient for that).
+Its main drawback is that a lot of temporaries may be created, and it is not entirely sure later optimizations will be able to get rid of them and to merge the loops. Even if it is, it may consume “optimization time and resource” to undo something that could have been done right the first time with only a little more work.
+
+This could be implemented as a single pass over the expression tree where the lowering for each expression node first asks for the lowering of its operands and expect to get a memory reference that materializes it (or an SSA value if the operand is scalar). In case the node being processed is an elemental operation, a `fir.loop` is generated to implement it. If a node is an array section or a call to a transformational, a box is created over the data reference or the result of the transformational argument expression evaluation.
+
+Short-circuits are generated also generated outside of any fir-loop which already is the case of other operands so nothing on top of generating the `fir.where` needs to be done.
+
+## Further Features that Expression Lowering
```

Comment:
I don't understand this title.  Did you mean to say "Further Features of Expression Lowering"?

Reply:
Oups, something like this, yes. It is now: "Possible Future Improvements of Array Expression Lowering"